File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UnnecessaryBitwiseOperation.Fixed.vb
Imports System
Imports System.Collections.Generic

Public Enum Types
    [Class] = 0
    Struct = 1
    [Public] = 2
    [Private] = 4
End Enum

Class UnnecessaryBitwiseOperation

    Public Sub Method()
        Dim Result As Integer
        Dim Zero As Integer = 0
        Dim One As Integer = 1
        Dim BitMask As Integer = &H10F

        Result = BitMask ' Fixed
        Result = BitMask   ' Fixed
        Result = BitMask  ' Fixed
        Result = BitMask  ' Fixed
        Result = BitMask   ' Fixed

        Result = BitMask And 1  ' Compliant
        Result = BitMask Or 1   ' Compliant
        Result = BitMask Xor 1  ' Compliant
        Result = BitMask Xor One    ' Compliant

        Dim BitMaskLong As Long = &H10F
        Dim ResultLong As Long
        ResultLong = BitMaskLong ' Fixed
        ResultLong = BitMaskLong And 0L     ' Compliant
        ResultLong = BitMaskLong     ' Fixed
        ResultLong = BitMaskLong    ' Fixed
        ResultLong = BitMaskLong And ReturnLong() ' Compliant
        ResultLong = BitMaskLong And &HF   ' Compliant

        Dim ResultULong As ULong = 1UL     ' Fixed
        ResultULong = 1UL Or 18446744073709551615UL ' Compliant

        MyMethod(1UL) ' Fixed

        If True AndAlso True Then Return
        If True OrElse True Then Return

        Dim Flags As Types = Types.Class Or Types.Private   ' Compliant even when class is zero
        Flags = Types.Class
        Flags = Flags Or Types.Private      ' Compliant, even when "flags" was initally zero
    End Sub

    Private Shared Function ReturnLong() As Long
        Return 1L
    End Function

    Private Shared Sub MyMethod(U As UInt64)
    End Sub

End Class

' https//github.com/SonarSource/sonar-dotnet/issues/4399
Public Class Repro_4399

    Public Sub BuildMask(DaysOfWeek As IEnumerable(Of DayOfWeek))
        Dim Value As Integer = 0
        For Each dow As DayOfWeek In DaysOfWeek
            Value = Value Or (1 << CInt(dow))   ' Compliant, value changes over iterations
        Next
    End Sub

    Public Sub Repro(Args() As Object)
        Dim Fail As Boolean = False
        For Each Arg As Object In Args
            Fail = Fail Or Not CheckArg(Arg)    ' Compliant, using short-circuit operator OrElse would change the logic.
        Next
    End Sub

    Private Function CheckArg(Arg As Object) As Boolean
    End Function

    Public Sub FindConstant_For_AssignedInsideLoop()
        Dim Value As Integer = 1
        Dim Result As Integer
        For v As Integer = 0 To 41
            Value = 0
            Result = v     ' Fixed
        Next
    End Sub

    Public Sub FindConstant_For_ReassignedToTheSameValue()
        Dim Value As Integer = 0
        Dim Result As Integer
        For v As Integer = 0 To 41
            Result = Value Or v     ' FN per rule description, but expected behavior for ConstantValueFinder. Variable "Value" is reassigned inside a Loop.
            Value = 0
        Next
    End Sub

    Public Sub FindConstant_For()
        Dim Value As Integer = 0
        Dim Unchanged As Integer = 0
        Dim Result As Integer
        For v As Integer = 0 To 41
            Result = Value Or v         ' Compliant, value changes over iterations
            Result = v     ' Fixed
            Value = 1
        Next
        Unchanged = 1
    End Sub

    Public Sub FindConstant_ForEach(Values() As Integer)
        Dim Value As Integer = 0
        Dim Unchanged As Integer = 0
        Dim Result As Integer
        For Each v As Integer In Values
            Result = Value Or v         ' Compliant, value changes over iterations
            Result = v     ' Fixed
            Value = 1
        Next
        Unchanged = 1
    End Sub

    Public Sub FindConstant_While(Values() As Integer)
        Dim Value As Integer = 0
        Dim Unchanged As Integer = 0
        Dim Result As Integer
        Dim Index As Integer
        While Index < Values.Length
            Dim v As Integer = Values(Index)
            Result = Value Or v         ' Compliant, value changes over iterations
            Result = v     ' Fixed
            Value = 1
            Index += 1
        End While
        Unchanged = 1
    End Sub

    Public Sub FindConstant_DoWhile(Values() As Integer)
        Dim Value As Integer = 0
        Dim Unchanged As Integer = 0
        Dim Result As Integer
        Dim Index As Integer
        Do While Index < Values.Length
            Dim v As Integer = Values(Index)
            Result = Value Or v         ' Compliant, value changes over iterations
            Result = v     ' Fixed
            Value = 1
            Index += 1
        Loop
        Unchanged = 1
    End Sub

    Public Sub FindConstant_DoUntil(Values() As Integer)
        Dim Value As Integer = 0
        Dim Unchanged As Integer = 0
        Dim Result As Integer
        Dim Index As Integer
        Do Until Index >= Values.Length
            Dim v As Integer = Values(Index)
            Result = Value Or v         ' Compliant, value changes over iterations
            Result = v     ' Fixed
            Value = 1
            Index += 1
        Loop
        Unchanged = 1
    End Sub

    Public Sub FindConstant_DoLoopWhile(Values() As Integer)
        Dim Value As Integer = 0
        Dim Unchanged As Integer = 0
        Dim Result As Integer
        Dim Index As Integer
        Do
            Dim v As Integer = Values(Index)
            Result = Value Or v         ' Compliant, value changes over iterations
            Result = v     ' Fixed
            Value = 1
            Index += 1
        Loop While Index < Values.Length
        Unchanged = 1
    End Sub

    Public Sub FindConstant_DoLoopUntil(Values() As Integer)
        Dim Value As Integer = 0
        Dim Unchanged As Integer = 0
        Dim Result As Integer
        Dim Index As Integer
        Do
            Dim v As Integer = Values(Index)
            Result = Value Or v         ' Compliant, value changes over iterations
            Result = v     ' Fixed
            Value = 1
            Index += 1
        Loop Until Index >= Values.Length
        Unchanged = 1
    End Sub

End Class

Public Class LocalFields

    Private Start As Integer = 0

    Public Property [End] As Integer = 0

    Public Sub UpdateStart(Val As Integer)
        Start = Val
    End Sub

    Public Overrides Function GetHashCode() As Integer
        Dim Value1 = Method() ^ [End]
        Dim Value2 = Start ^ [End]
        Dim Value3 = [End] ^ Start
        Return Method() ^ Start
    End Function

    Private Function Method() As Integer
        Return 42
    End Function

End Class

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UnnecessaryBitwiseOperation.Fixed.cs
using System;
using System.Collections.Generic;

namespace Tests.Diagnostics
{
    public enum Types
    {
        Class = 0,
        Struct = 1,
        Public = 2,
        Private = 4
    }

    class UnnecessaryBitwiseOperation
    {
        static void Main(string[] args)
        {
            int result;
            int bitMask = 0x010F;

            result = bitMask; // Fixed
            result = bitMask;  // Fixed
            result = bitMask;  // Fixed
            result = bitMask;  // Fixed
            var result2 = result;  // Fixed

            result = bitMask & 1; // Compliant
            result = bitMask | 1; // compliant
            result = bitMask ^ 1; // Compliant
            result &= 1; // Compliant
            result |= 1; // compliant
            result ^= 1; // Compliant

            long bitMaskLong = 0x010F;
            long resultLong;
            resultLong = bitMaskLong; // Fixed
            resultLong = bitMaskLong & 0L; // Compliant
            resultLong = bitMaskLong; // Fixed
            resultLong = bitMaskLong; // Fixed
            resultLong = bitMaskLong & returnLong(); // Compliant
            resultLong = bitMaskLong & 0x0F; // Compliant

            var resultULong = 1UL; // Fixed
            resultULong = 1UL | 18446744073709551615UL; // Compliant

            MyMethod(1UL); // Fixed

            var flags = Types.Class | Types.Private; // Compliant even when Class is zero
            flags = Types.Class;
            flags = flags | Types.Private;  // Compliant, even when "flags" was initally zero
        }

        public static void WithUnaryPrefix()
        {
            int result;
            int bitMask = 0x010F;
            int one = 1;
            int zero = 0;

            result = bitMask & -one;       // FN - Unary Operator is not supported
            result &= -one;                // FN - Unary Operator is not supported
            result = bitMask & - - -+one;  // FN - Unary Operator is not supported
            result = bitMask | + + +zero;  // FN - Unary Operator is not supported
            result = bitMask | + + +zero;  // FN - Unary Operator is not supported
        }

        public void UnaryOperators()
        {
            var length = 0x80;

            var bytes1 = 0;
            bytes1++;
            _ = bytes1 | 0x80; // Compliant, See: https://github.com/SonarSource/sonar-dotnet/issues/6326

            var bytes2 = 0;
            ++bytes2;
            _ = bytes2 | 0x80; // Compliant

            var bytes3 = 0;
            bytes3--;
            _ = bytes3 | 0x80; // Compliant

            var bytes4 = 0;
            --bytes4;
            _ = bytes4 | 0x80; // Compliant

            var bytes5 = 0;
            var bytesOther = 0;
            --bytesOther;      // Other variable is mutated
            _ = 0x80; // Fixed

            var bytes6 = 0;
            bytesOther++;      // Other variable is mutated
            _ = 0x80; // Fixed
        }

        private static long returnLong()
        {
            return 1L;
        }

        private static void MyMethod(UInt64 u) { }
    }
}

// https://github.com/SonarSource/sonar-dotnet/issues/4399
public class Repro_4399
{
    public void BuildMask(IEnumerable<DayOfWeek> daysOfWeek)
    {
        var value = 0;
        foreach (var dow in daysOfWeek)
        {
            value = value | (1 << (int)dow); // Compliant, value changes over iterations
        }
    }

    public void Repro(object[] args)
    {
        var fail = false;
        foreach (var arg in args)
        {
            fail = fail | !CheckArg(arg);   // Compliant, using short-circuit operator || would change the logic.
        }
    }

    private bool CheckArg(object arg) => false;

    public void FindConstant_For_AssignedInsideLoop()
    {
        var value = 1;
        int result;
        for (var v = 0; v < 42; v++)
        {
            value = 0;
            result = v;     // Fixed
        }
    }

    public void FindConstant_For_ReassignedToTheSameValue()
    {
        var value = 0;
        int result;
        for (var v = 0; v < 42; v++)
        {
            result = value | v;     // FN per rule description, but expected behavior for ConstantValueFinder. Variable "value" is reassigned inside a loop.
            value = 0;
        }
    }

    public void FindConstant_For()
    {
        var value = 0;
        var unchanged = 0;
        int result;
        for (var v = 0; v < 42; v++)
        {
            result = value | v;     // Compliant, value changes over iterations
            result = v; // Fixed
            value = 1;
        }
    }

    public void FindConstant_ForEach(int[] values)
    {
        var value = 0;
        var unchanged = 0;
        int result;
        foreach (var v in values)
        {
            result = value | v;     // Compliant, value changes over iterations
            result = v; // Fixed
            value = 1;
        }
        unchanged = 1;
    }

    public void FindConstant_While(int[] values)
    {
        var value = 0;
        var unchanged = 0;
        int result;
        int index = 0;
        while (index < values.Length)
        {
            var v = values[index];
            result = value | v;     // Compliant, value changes over iterations
            result = v; // Fixed
            value = 1;
            index++;
        }
        unchanged = 1;
    }

    public void FindConstant_Do(int[] values)
    {
        var value = 0;
        var unchanged = 0;
        int result;
        var index = 0;
        do
        {
            var v = values[index];
            result = value | v;     // Compliant, value changes over iterations
            result = v; // Fixed
            value = 1;
            index++;
        } while (index < values.Length);
        unchanged = 1;
    }

    class LocalFields
    {
        private int start = 0;

        public int End { get; set; } = 0;

        public void UpdateStart(int val) => start = val;

        public override int GetHashCode()
        {
            var value1 = Method() ^ End;
            var value2 = start ^ End;
            var value3 = End ^ start;
            return Method() ^ start;

            int Method() => 42;
        }
    }
}
